spec/requests/case_logs_controller_spec.rb

```diff
@@ -34,84 +34,104 @@
+        expect(json_response["age1"]).to eq(age1)
+        expect(json_response["property_postcode"]).to eq(property_postcode)
       end
-    end
 
-    context "with a complete case log submission" do
-      let(:org_params) do
-        {
-          "case_log" => {
-            "owning_organisation_id" => owning_organisation.id,
-            "managing_organisation_id" => managing_organisation.id,
-          },
-        }
+      context "with invalid json parameters" do
+        let(:age1) { 2000 }
+        let(:offered) { 21 }
+
+        it "validates case log parameters" do
+          json_response = JSON.parse(response.body)
+          expect(response).to have_http_status(:unprocessable_entity)
+          expect(json_response["errors"]).to match_array([["offered", [I18n.t("validations.property.offered.relet_number")]], ["age1", [I18n.t("validations.household.age.must_be_valid", lower_bound: 16)]]])
+        end
       end
-      let(:case_log_params) { JSON.parse(File.open("spec/fixtures/complete_case_log.json").read) }
-      let(:params) do
-        case_log_params.merge(org_params) { |_k, a_val, b_val| a_val.merge(b_val) }
+
+      context "with a partial case log submission" do
+        it "marks the record as in_progress" do
+          json_response = JSON.parse(response.body)
+          expect(json_response["status"]).to eq(in_progress)
+        end
       end
 
-      it "marks the record as completed" do
-        json_response = JSON.parse(response.body)
+      context "with a complete case log submission" do
+        let(:org_params) do
+          {
+            "case_log" => {
+              "owning_organisation_id" => owning_organisation.id,
+              "managing_organisation_id" => managing_organisation.id,
+            },
+          }
+        end
+        let(:case_log_params) { JSON.parse(File.open("spec/fixtures/complete_case_log.json").read) }
+        let(:params) do
+          case_log_params.merge(org_params) { |_k, a_val, b_val| a_val.merge(b_val) }
+        end
+
+        it "marks the record as completed" do
+          json_response = JSON.parse(response.body)
+
+          expect(json_response).not_to have_key("errors")
+          expect(json_response["status"]).to eq(completed)
+        end
+      end
+
+      context "with a request containing invalid credentials" do
+        let(:basic_credentials) do
+          ActionController::HttpAuthentication::Basic.encode_credentials(api_username, "Oops")
+        end
 
-        expect(json_response).not_to have_key("errors")
-        expect(json_response["status"]).to eq(completed)
+        it "returns 401" do
+          expect(response).to have_http_status(:unauthorized)
+        end
       end
     end
 
-    context "with a request containing invalid credentials" do
-      let(:basic_credentials) do
-        ActionController::HttpAuthentication::Basic.encode_credentials(api_username, "Oops")
+    context "when UI" do
```

Comment:
Not a big deal but I think we should export our UI tests outside this file at some point (not now).
The "when API" and "when UI" are not really contexts, more how we test the application (at least to me). But I like that it logically splits the code which would make a refactor easier.

Reply:
They kind of are contexts (I could also have said: "when request header is json" and "when request header is html") but happy to refactor. Will leave that out of scope for this PR though.